src/Features/Core/Portable/EditAndContinue/EditSession.cs

```diff
@@ -4,116 +4,145 @@
+            Debug.Assert(telemetry != null);
 
-            _baseSolution = baseSolution;
-            _debuggingSession = debuggingSession;
-            _activeStatementProvider = activeStatementProvider;
-            _stoppedAtException = stoppedAtException;
-            _projects = projects;
-            _cancellation = new CancellationTokenSource();
+            DebuggingSession = debuggingSession;
 
-            // TODO: small dict, pool?
-            _analyses = new Dictionary<DocumentId, Analysis>();
+            _analyses = new Dictionary<DocumentId, (Document Document, AsyncLazy<DocumentAnalysisResults> Results)>();
+            _documentsWithReportedDiagnostics = new HashSet<DocumentId>();
+            _moduleDiagnostics = new Dictionary<Guid, ImmutableArray<LocationlessDiagnostic>>();
 
-            // TODO: small dict, pool?
-            _documentsWithReportedRudeEdits = new HashSet<DocumentId>();
+            _nonRemappableRegions = debuggingSession.NonRemappableRegions;
 
-            _nonRemappableRegions = nonRemappableRegions;
+            Telemetry = telemetry;
+            BaseSolution = debuggingSession.LastCommittedSolution;
 
             BaseActiveStatements = new AsyncLazy<ActiveStatementsMap>(GetBaseActiveStatementsAsync, cacheResult: true);
             BaseActiveExceptionRegions = new AsyncLazy<ImmutableArray<ActiveStatementExceptionRegions>>(GetBaseActiveExceptionRegionsAsync, cacheResult: true);
+            StoppedAtException = stoppedAtException;
+            Cancellation = new CancellationTokenSource();
+        }
+
+        internal void ModuleInstanceLoadedOrUnloaded(Guid mvid)
+        {
+            // invalidate diagnostic cache for the module:
+            lock (_moduleDiagnosticsGuard)
+            {
+                _moduleDiagnostics.Remove(mvid);
+            }
+        }
+
+        public ImmutableArray<LocationlessDiagnostic> GetModuleDiagnostics(Guid mvid, string projectDisplayName)
+        {
+            ImmutableArray<LocationlessDiagnostic> result;
+            lock (_moduleDiagnosticsGuard)
+            {
+                if (_moduleDiagnostics.TryGetValue(mvid, out result))
+                {
+                    return result;
+                }
+            }
+
+            var newResult = ImmutableArray<LocationlessDiagnostic>.Empty;
+            if (!DebuggingSession.DebugeeModuleMetadataProvider.IsEditAndContinueAvailable(mvid, out var errorCode, out var localizedMessage))
+            {
+                var descriptor = EditAndContinueDiagnosticDescriptors.GetModuleDiagnosticDescriptor(errorCode);
+                newResult = ImmutableArray.Create(new LocationlessDiagnostic(descriptor, new[] { projectDisplayName, localizedMessage }));
+            }
+
+            lock (_moduleDiagnosticsGuard)
+            {
+                if (!_moduleDiagnostics.TryGetValue(mvid, out result))
+                {
+                    _moduleDiagnostics.Add(mvid, result = newResult);
```

Comment:
💡 I believe we have an extension method `IDictionaryExtensions.GetOrAdd` for this

Reply:
We do, but it takes a delegate and thus we would need to allocate closure unnecessarily.

---
In reply to: [270626551](https://github.com/dotnet/roslyn/pull/34027#discussion_r270626551) [](ancestors = 270626551)